apps/vmq_diversity/src/vmq_diversity_worker_wrapper.erl

```diff
@@ -0,0 +1,191 @@
+%% Initializes the server
+%%
+%% @spec init(Args) -> {ok, State} |
+%%                     {ok, State, Timeout} |
+%%                     ignore |
+%%                     {stop, Reason}
+%% @end
+%%--------------------------------------------------------------------
+init(Opts) ->
+    process_flag(trap_exit, true),
+    self() ! connect,
+    {M,F,WorkerArgs} = proplists:get_value(start_mfa, Opts),
+    ReconnectTimetout = proplists:get_value(reconnect_timeout, Opts, 1000),
+    {ok, #state{start_mfa = {M,F,WorkerArgs},
+                reconnect_timeout = ReconnectTimetout}}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling call messages
+%%
+%% @spec handle_call(Request, From, State) ->
+%%                                   {reply, Reply, State} |
+%%                                   {reply, Reply, State, Timeout} |
+%%                                   {noreply, State} |
+%%                                   {noreply, State, Timeout} |
+%%                                   {stop, Reason, Reply, State} |
+%%                                   {stop, Reason, State}
+%% @end
+%%--------------------------------------------------------------------
+handle_call({apply, _,_,_}, _From, #state{worker=undefined}=State) ->
+    {reply, {error, not_connected}, State};
+handle_call({apply, Mod, Fun, Args}, _From, #state{worker={_MRef, Pid}}=State) when is_pid(Pid) ->
+    Reply =
+        try
+            erlang:apply(Mod, Fun, [Pid|Args])
+        catch
+            exit:{noproc,_Reason} ->
+                %% We remove the worker pid only in handle_info where
+                %% the attempt to restart the process will also be
+                %% kicked off.
+                {error, not_connected}
+        end,
+    {reply, Reply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling cast messages
+%%
+%% @spec handle_cast(Msg, State) -> {noreply, State} |
+%%                                  {noreply, State, Timeout} |
+%%                                  {stop, Reason, State}
+%% @end
+%%--------------------------------------------------------------------
+handle_cast(_Msg, State) ->
+    {noreply, State}.
+
+%%--------------------------------------------------------------------
+%% @private
+%% @doc
+%% Handling all non call/cast messages
+%%
+%% @spec handle_info(Info, State) -> {noreply, State} |
+%%                                   {noreply, State, Timeout} |
+%%                                   {stop, Reason, State}
+%% @end
+%%--------------------------------------------------------------------
+handle_info(connect, #state{start_mfa={M,F,A}, worker=undefined} = State) ->
+    NewState =
+    case apply(M,F,[A]) of
+        {ok, Pid} ->
+            MRef = monitor(process, Pid),
+            State#state{worker={MRef, Pid}};
+        {error, Reason} ->
+            lager:warning("Could not connect due to ~p", [Reason]),
```

Comment:
Let's add some log information containing the `M` and maybe `F`... so one knows which underlying connection is crashing

Reply:
I'll change the start_mfa to pass in a fun/0 instead and then also add a name we can use here for logging